### PR TITLE
bump-formula-pr: determine url/tag from version

### PR DESCRIPTION
**Pull request by SeekingMeaning**
- [ ] Have you followed the guidelines in our [Contributing](https://github.com/Homebrew/brew/blob/HEAD/CONTRIBUTING.md) document?
- [ ] Have you checked to ensure there aren't other open [Pull Requests](https://github.com/Homebrew/brew/pulls) for the same change?
- [ ] Have you added an explanation of what your changes do and why you'd like us to include them?
- [ ] Have you written new tests for your changes? [Here's an example](https://github.com/Homebrew/brew/blob/HEAD/Library/Homebrew/test/PATH_spec.rb).
- [ ] Have you successfully run `brew style` with your changes locally?
- [ ] Have you successfully run `brew tests` with your changes locally?

-----

### Examples:

Before this PR:

```bash
brew bump-formula-pr sops --url https://github.com/mozilla/sops/archive/v3.6.0.tar.gz

brew bump-formula-pr kustomize --tag kustomize/v3.8.1 --revision 0b359d0ef0272e6545eda0e99aacd63aef99c4d0
```

After this PR:

```bash
brew bump-formula-pr sops --version 3.6.0				# Determines new URL and SHA-256

brew bump-formula-pr kustomize --version 3.8.1			# Determines new tag and revision
brew bump-formula-pr kustomize --tag kustomize/v3.8.1	# Determines new revision only
```

What the patches would look like:

```patch
-  url "https://github.com/mozilla/sops/archive/v3.5.0.tar.gz"
-  sha256 "a9c257dc5ddaab736dce08b8c5b1f00e6ca1e3171909b6d7385689044ebe759b"
+  url "https://github.com/mozilla/sops/archive/v3.6.0.tar.gz"
+  sha256 "fee1c27f14f9f45b5955627e301aafcc38973c9458b25f99ef241bdd0a3b082c"
```

```patch
   url "https://github.com/kubernetes-sigs/kustomize.git",
-      :tag      => "kustomize/v3.8.0",
-      :revision => "6a50372dd5686df22750b0c729adaf369fbf193c"
+      :tag      => "kustomize/v3.8.1",
+      :revision => "0b359d0ef0272e6545eda0e99aacd63aef99c4d0"
```

Links to formulae:

* [`sops.rb`](https://github.com/Homebrew/homebrew-core/blob/master/Formula/sops.rb)
* [`kustomize.rb`](https://github.com/Homebrew/homebrew-core/blob/master/Formula/kustomize.rb)

-----

### Decision chart:

Decision chart

-----

Added `determine_formula_from_url()` to fix style offenses:

```
C: Metrics/AbcSize: Assignment Branch Condition size for bump_formula_pr is too high. [<88, 240, 104> 276/275]
  def bump_formula_pr ...
  ^^^^^^^^^^^^^^^^^^^
C: Metrics/CyclomaticComplexity: Cyclomatic complexity for bump_formula_pr is too high. [92/85]
  def bump_formula_pr ...
  ^^^^^^^^^^^^^^^^^^^
```

### CODE REVIEW

**Comment by jonchang** [comment]
Can you write some examples for what we would be able to do with `brew bump-formula-pr` with this pull request that we wouldn't be able to do before?

**Comment by SeekingMeaning** [comment]
Done! Added examples to the description

The commands that we'll be to use will look like this:

```bash
brew bump-formula-pr sops --version 3.6.0			# Determines new URL and SHA-256
brew bump-formula-pr kustomize --version 3.8.1		# Determines new tag and revision
```

I've been using this branch for the past day or so and tried to fix any unexpected behavior

**Comment by dawidd6** [comment]
I'm very much in favor of this. Making those changes could enable us to easily integrate the bumping procedure with livecheck and automate it even further in the future.

**Comment by jonchang** [comment]
Merging this, thanks for the new feature and explanation! Also @SeekingMeaning please check the email listed on your GitHub profile, thanks!

File: Library/Homebrew/dev-cmd/bump-formula-pr.rb [hunk 1]
```diff
@@ -175,28 +155,41 @@ def bump_formula_pr
...
       false
     elsif !hash_type
-      odie "#{formula}: no --tag=/--revision= arguments specified!"
-    elsif !new_url
-      odie "#{formula}: no --url= argument specified!"
+      odie "#{formula}: no --tag= or --version= argument specified!" if !new_tag && !new_version
+      new_tag ||= old_tag.gsub(old_version, new_version)
+      if new_tag == old_tag
+        odie <<~EOS
+          You probably need to bump this formula manually since the new tag
```
Comment (MikeMcQuaid) [inline Library/Homebrew/dev-cmd/bump-formula-pr.rb:173]: Can we drop the `probably` here? It doesn't make it clear to the contributor what they should do.

File: Library/Homebrew/dev-cmd/bump-formula-pr.rb [hunk 2]
```diff
@@ -175,28 +155,41 @@ def bump_formula_pr
...
-      if forced_version
-        if forced_version == resource.version
-          forced_version = nil
-        else
-          resource.version = forced_version
-        end
+      new_url ||= old_url.gsub(old_version, new_version)
+      if new_url == old_url
+        odie <<~EOS
+          You probably need to bump this formula manually since the new URL
```
Comment (MikeMcQuaid) [inline Library/Homebrew/dev-cmd/bump-formula-pr.rb:187]: Can we drop the `probably` here? It doesn't make it clear to the contributor what they should do.

File: Library/Homebrew/dev-cmd/bump-formula-pr.rb [hunk 3]
```diff
@@ -413,6 +404,40 @@ def bump_formula_pr
...
+    guesses = []
+    Formula.each do |f|
+      if is_devel && f.devel && f.devel.url && f.devel.url.match(base_url)
+        guesses << f
+      elsif f.stable&.url && f.stable.url.match(base_url)
+        guesses << f
+      end
+    end
+    return guesses.shift if guesses.count == 1
+    return unless guesses.count > 1
```
Comment (MikeMcQuaid) [inline Library/Homebrew/dev-cmd/bump-formula-pr.rb:426]: `return if guesses.count <= 1`

File: Library/Homebrew/dev-cmd/bump-formula-pr.rb [hunk 4]
```diff
@@ -413,6 +404,40 @@ def bump_formula_pr
...
+    # first five components since sometimes the last component isn't the only
+    # one to change.
+    url_split = url.split("/")
+    maximum_url_components_to_match = 5
+    components_to_match = [url_split.count - 1, maximum_url_components_to_match].min
+    base_url = url_split.first(components_to_match).join("/")
+    base_url = /#{Regexp.escape(base_url)}/
+    guesses = []
+    Formula.each do |f|
+      if is_devel && f.devel && f.devel.url && f.devel.url.match(base_url)
```
Comment (MikeMcQuaid) [inline Library/Homebrew/dev-cmd/bump-formula-pr.rb:419]: I think we can remove all `devel` logic now as it's been deprecated,

**Pull request merged**